### ISSUE TITLE
investigate 'kakdiff'

### ISSUE THREAD

**Issue by alexherbo2** [mawww/kakoune#153]

**Comment by dumblob** [comment]
+1

For reference, the current Vim uses a "dumb" solution - running `diff -urN` in background and opening the result in two buffers with *vsplit* between them and *foldenable* in each of them; guys from https://github.com/neovim/neovim are investigating how to make it better in terms of speed, safety, removal of the need to manually run `:diffupdate` etc. - see https://github.com/neovim/neovim/issues/1007 .

**Comment by lenormf** [comment]
I think this issue can be closed, as it's not something that would be merged into the core, but rather a script.

**Comment by dumblob** [comment]
@lenormf but it shall be provided by default with the `kak` distribution, so I think it should rather stay opened. Btw this issue was also for quite long in the kakoune TODO list, which got transformed into separate Issues on GitHub and could have been removed. So it seems to me it was even always meant to be a part of the default `kak` distribution.

**Comment by lenormf** [comment]
What exactly would such script do? It seems that it would only be a wrapper of the `diff` command with some default arguments, I'm not sure how doing this from within the editor would be more convenient than from a shell.

`vim` and its forks (`emacs` too) like to integrate all the tools that compliment a code editor in general just for the sake of "doing thinks from `$EDITOR`", but that doesn't fit the minimalist paradigm that kakoune uses. Unless there are features that would make using `diff` really easier from within kakoune, there's no real point keeping this open as it wouldn't be a very relevant addition to the main repository. Somebody with a personal `extra` repository might have implemented it already, or maybe the script should go to the wiki, I don't know.

**Comment by dumblob** [comment]
>What exactly would such script do? It seems that it would only be a wrapper of the diff command with some default arguments, I'm not sure how doing this from within the editor would be more convenient than from a shell.

`kakdiff`, `vimdiff` , `*diff` are **editable** variants of `diff` - editing is the ultimate added value. It's the tool number one for anyone working with versioning systems (git, svn, csv, mercurial, darcs, many others). It's an indispensable tool for any update (especially in power-user focused systems like most unixes and Linux distributions - think of configuration files updates related to packaging, configuring of anything as unix-like systems have human-centric configuration in text files, etc.). There are many other specific use cases, but these are probably the most prominent ones.

IMHO `kakdiff` can be implemented without touching C++ in the `kak` core (i.e. while keeping the overall minimalism of `kak`), but I don't want to let this issue be closed because a *diff editing* functionality is undoubtedly an **essential** part of everyday editing for which one usually uses an *editor*.

**Comment by alexherbo2** [comment]
A diff tool that can embed editors by connecting to them would be awesome.

**Comment by alexherbo2** [comment]
@lenormf I’ll like to keep it open until a solution is provided

**Comment by lenormf** [comment]
>a diff editing functionality is undoubtedly an essential part of everyday editing for which one usually uses an editor

I still fail to see how that's a critical feature, what is a specific example of a situation where I would like to manually edit a patch?

**Comment by dumblob** [comment]
>I still fail to see how that's a critical feature, what is a specific example of a situation where I would like to manually edit a patch?

First, what I wrote (and also meant) is *diff editing* functionality, **not** just patch editing functionality. I strongly believe you're running *diff* not just for fun, but to see what are the differences to **act** upon the findings (yeah, acting means in 99.99% cases editing).

Second, you end up with *patch editing* also quite often (because of merge conflicts as no VCS workflow is error-less; because all SW devs who want to have a high quality code do patch reviews to others; etc.). And yes, therefore even this *patch editing* is critical.

**Comment by lenormf** [comment]
>diff editing functionality, not just patch editing functionality

A patch is also called a "diff" because of the name of the `diff` utility which generates patches. With that in mind, if you wanted to edit a patch you could simply use the `!` key in kakoune to insert the output of `diff` (or e.g. `git diff`), or simply open a saved .patch file, and edit it. Which means that a `kakdiff` script would be a couple lines long, which isn't worth merging upstream, making this issue clearly superficial as people who need such a script have probably implemented it themselves already.

Unless I'm missing something about `vimdiff` that differenciates it from a glorified alias, which is why I asked about an example of a situation where it's needed.

**Comment by dumblob** [comment]
>Which means that a kakdiff script would be a couple lines long, which isn't worth merging upstream, making this issue clearly superficial as people who need such a script have probably implemented it themselves already.

Unfortunately not - I looked into it and it isn't by far that simple. So I'm pretty confident people didn't implement it yet. Based on that I'm still convinced it should be part of the default `kak` distribution.

If you want to prove me/us wrong, please stick with the famous phrase `Talk is cheap. Show me the code.`.

**Comment by lenormf** [comment]
>I looked into it and it isn't by far that simple

I thought my last messages made it clear that I'm still unsure what more there is to it than simply showing the output of a shell command, so I'll spell it out this time: what does `vimdiff` do that makes it better than passing the usual UNIX tools to `!` ?

I noticed that `vimdiff` provides with some bindings, which funnily enough I do have a few examples of in my own configuration:

```
## jump to the next hunk using the `@` key in diff buffers
## jump to the next file with the `#` key
hook global WinSetOption filetype=diff %{
    map window normal '@' /^@@<ret>
    map window normal '#' /^diff\><ret>
}
```

**Comment by cbuschardt** [comment]
The request here is to provide a side-by-side diff that can be edited in real-time.  This is wildly different than diff output.  Consider: Editable, Syntax coloring, movable split, etc.  

How does a script create a vertical split reliably [i3wm/gnome/tmux]?

**Comment by dumblob** [comment]
>what does vimdiff do that makes it better than passing the usual UNIX tools to `!` ?

Please refer to http://vimdoc.sourceforge.net/htmldoc/diff.html (and read it very carefully) and then look around the internet for major use cases like [merging](https://gist.github.com/mattratleph/4026987 ) to get an idea how it's actually used. Then implement majority, if not all, of the vimdiff functionality and then please post the code here. You can start with one-by-one implementation of all vimdiff commands from the linked vimdoc.

I won't blame you if you give up with the implementation (because as I already wrote, it's not at all as easy as it seems), but please, just post any your results (even if none) to prove, you've understood what is vimdiff about. I'm encouraging you here as I still can't believe at all you missed the very most important features of vimdiff (this observation is among other things based on the code you've posted as it seems like a really bad joke in the context of vimdiff and our discussion).

**Comment by andreyorst** [comment]
When working with a project that may have relatively similar files (by design) the side-by-side diff tool in the editor is **very** useful. So this isn't really every time about VCS diff or patch editing.

**Comment by mawww** [comment]
@lenormf I have not used kakmerge yet, but wouldnt it almost cater for the diff use case as well ? Or is that pretty different ?

**Comment by xxxserxxx** [comment]
I'd like to add some flavor to @lenormf's last comment.

I use diff outside of VCS pretty regularly.  As an example, SyncThing produces conflict files, basically a backup copy of a file before it overwrites it.  When these happen, and they're on text files, I'd like to be able to diff and manually merge them -- there's no VCS in the picture.

As a second use case, my VCS preference is Mercurial; I only use git when I'm forced to.  I really hope that a monoculture where git is the only option isn't something that we're considering promoting -- I don't know if kakmerge depends on git, but all of the documentation for it implies that it is tightly coupled to git  -- the github short description is, literally, "A Kakoune-based mergetool for Git."  

TL;DR: I hope we can see a solution that does not *only* support, and is not tightly coupled to, a tool who's primary function is not diffing.

**Comment by andreyorst** [comment]
I also often diff current buffer with file, in case I've forgot what I've changed, or I've used some formatter and want to quickly check what was changed by it.

**Comment by tmccombs** [comment]
To throw another use case out there, I currently use vims diff mode to compare old and new versions of config after upgrading software (pacnew files on archlinux), and accept changes I want. Also, when working with git diffs, I generally prefer side by side diffs. With fugitive, I can stage, unstage, or restore chunks from a side-by-side diff.

**Comment by tmccombs** [comment]
Also, it looks like kakmerge requires either tmux or kitty. What if you use a different solution, like terminator, iterm2, i3, screen, etc.

**Comment by andreyorst** [comment]
I believe that we should focus on a replace-ranges approach. Side by side diff is the best thing, but making this in Kakoune is notoriously hard, because we will need to implement synchronisation of different windows, so those scroll in sync, and line on the left is always aligned to the one in the right. This may not be that hard in tmux, because we can specify that we want two side by side windows of the same height, but can't really be done for window managers (I think).

If we will stick to replace ranges approach, we will be able to insert both variants inside current buffer, without modifying the buffer itself, and add highlighting for different variants.

**Comment by tmccombs** [comment]
I'm not sure what you mean by "replace-ranges approach"

**Comment by andreyorst** [comment]
parse diff tool output and add replace ranges highlighters to buffer that will show parsed diff chunks

**Comment by tmccombs** [comment]
If I understand correctly, that would mean you can only change one side of the diff.  Since you don't have a buffer for the otherside. That may be fine for some use cases. But in some cases, I want to be able to make changes to both sides.

**Comment by xxxserxxx** [comment]
@andreyorst , is most of the complexity in the synchronization?  E.g., if you can assume tmux, is the implementation straightforward, or is it still complex? I'm wondering if there's a graceful-fallback opportunity here; it's trivial to detect whether kakoune is run in tmux (e.g. `[[ -n $TMUX ]]`).

**Comment by andreyorst** [comment]
> @andreyorst , is most of the complexity in the synchronization? E.g., if you can assume tmux, is the implementation straightforward, or is it still complex? I'm wondering if there's a graceful-fallback opportunity here; it's trivial to detect whether kakoune is run in tmux (e.g. `[[ -n $TMUX ]]`).

Kakoune can detect that we're in Tmux, and it already does this, by defining aliases to `terminal` command. Synchronisation is a hard thing not because we have to ensure that both diff chunks are on screen, and aligned, so user could really see the difference. There are many ways to attempt it e.g. in VSCode scrolling is completely mirrored between buffers. Emacs' ediff **only** allows jumping between chunks, and jumps synchronize cursor positions in buffers. Given that Kakoune has view mode that can be used for this, I don't think that it would  be major problem to achieve.

The real problem I think is split management itself. Yes, we can detect that we're inside tmux. But can you really predict how to create split that will work out for side by side comparison in this setup if I were to view diff of any of the buffers:

![image](https://user-images.githubusercontent.com/19470159/83439915-ae7b3000-a44c-11ea-96a2-74ed2521bcf6.png)

And even though it might be possible, because Tmux can be talked to, and is a fairly common choice along Kakoune users, there are too many other ways to split view, kitty, i3, bspwm, awesome, GNOME, any other X/Wayland/Mir/watever wm, iterm, GNU Screen, which all need to be supported for side by side diff. And we're not talking about three way diff yet, which is whole another level of synchronisation and alignment of windows in all of these cases.

So realistically I feel that single-buffer approach can be supported really well by Kakoune in it's current form, and side by side diff is too complex task and should be lifted as a separate plugin.

**Comment by tmccombs** [comment]
> parse diff tool output and add replace ranges highlighters to buffer that will show parsed diff chunks

> I feel that single-buffer approach can be supported really well by Kakoune in it's current form

I'm not sure how that would work. You can't select the display text (or at least it isn't highlighted,  (you can select the original text, but it isn't highlighted at all), I haven't been able to find a way to insert new lines of replaced text, and I have no idea how editing the other side of the diff would work.

I'd be willing build a plugin for this, but I'm not sure how to go about it.

**Comment by tmccombs** [comment]
It also seems like `flag-lines` highlights _almost_ works (and gives a side-by-side diff). Except, again you can't add new lines in the flag lines.

**Comment by andreyorst** [comment]
I also think that setting correct `flag-lines` width can be a problem

**Comment by xxxserxxx** [comment]
@andreyorst 

Thanks for the explanation of the challenges involved in the synchronization.

> even though it might be possible, because Tmux can be talked to, and is a fairly common choice along Kakoune users, there are too many other ways to split view, kitty, i3, bspwm, awesome, GNOME, any other X/Wayland/Mir/watever wm, iterm, GNU Screen, which all need to be supported for side by side diff.

What I was suggesting was a graceful fallback.  The default is the highlights approach, which works everywhere, but if tmux is available -- and it *is* common, has good support, everybody can run it without changing their favorite terminal or window manager, and so it makes a reasonable target -- then a user gets a much more functional and useful side-by-side diff. Or, in other words, if you want it fancy, you have to use tmux; otherwise, you get bare bones.  It's like `picom` -- some features depend on additional software.

What it sounds like is a drive to a lowest common denominator is preferred. Or maybe it's just too hacky to have two behaviors, chosen by which environment you're running in.  In either case, I wanted to clarify what I was suggesting: IF tmux THEN fully functional ELSE basic.

**Comment by Parasrah** [comment]
@xxxserxxx just wanted to provide the viewpoint of someone who specifically does not use `tmux` with kakoune, which I agree is probably a small portion of the overall community, that choice is very intentional and I would sooner go back to neovim than start using `tmux`. I liked kakoune because it works seamlessly (albeit with a bit of hacking) with everything. I use `i3`, and one of the highlights for me is now I can always use the same keys to move to another window, regardless of whether that window has `kakoune` running in it or a browser, and I have no interest in going back to having different modes with tmux.

Again I recognize I'm in the minority in this case, but after speaking with people about `kakoune` it's apparent that diff support is one of the main barriers to adoption coming from vim. And selling to those people (myself included) that they must use `tmux` to get half decent diff support might be harder than you think. If the official solution to this problem ends up being coupled tightly to `tmux`, I think it's going to take incentive and resources away from a simpler solution that could potentially work for everyone.

I like the suggestion of thinking outside the box of how it's conventionally been done with splits, but if diffsplit ends up being vital I'd at least like to see it done in a way that is window manager agnostic, with batteries for some of the easier ones (like tmux and kitty), so the rest of us can at least hack together something that works for us.

That's just my two cents though, and I'm pretty new to the community so I might be lacking context

**Comment by xxxserxxx** [comment]
@Parasrah I sympathize; I've been using dvtm instead of tmux for the past couple of weeks myself.

We have **no** visual diff support in kakoune right now, so non-tmux users are not losing anything if kakoune added support for visual diffs only in tmux.  If kakoune added visual diffs through tmux, you'd be no worse (or better) off than you are right now.  At least tmux is available for multiple OSes.

There's no standard for IPC at this level for window managers or terminal emulators: no common communication protocol to tell a WM to switch windows or identify layouts, or to tell a terminal emulator to switch tabs.  Supporting multiple protocols is also a nightmare -- maybe *someone* will want to support that, but better them than me.  In the meantime, I'd be happy to use tmux to get visual diffs -- it'd **still** be better than what I have to do now, which is to switch to vim to do them, and the tmux modes are far less divergent from i3 than the vim editing metaphors are from kakoune.

In any case, this would probably be better as a plugin, although the kak file would be pretty gnarly, and far beyond my kak scripting skills.

Just my $0.02

**Comment by Parasrah** [comment]
@xxxserxxx I agree with most of that, and see your point that if a better user experience could be achieved via `tmux` than without, it would likely benefit the majority of `kakoune` users, and about potential pain points of alternative approaches. I do think that non-tmux users do lose something if there is a great experience tied to `tmux`, because there is less incentive to improve the experience for non-tmux users. Again, that might well be a worthy trade-off, I have no idea. Personally it doesn't make a significant difference to me, I don't spend much time dealing with git conflicts anymore. I just know it is a critical feature for some.

Sorry if my previous comment came off confrontational or unpleasant, that wasn't my intention at all (unsure if that's what the :eyes: conveys)

**Comment by mawww** [comment]
I dont think we necessarily need to tie such a feature to tmux, all we want is a way to create a side by side window to display both versions, and sync scrolling.

To create a side by side window we just need a variant of the `terminal` alias where we can express that we want to split vertically (or is it horizontally ?). We can then implement that in support scripts for tmux, iterm2, kitty... (actually this is already implemented in those). For the general x11 support we cannot guarantee that split, but we can still just create the terminal and let the user take care of positionning. We can also write a i3.kak and dvtm.kak to use those WMs IPC support to split correctly.

All that to say, a side by side diff script does not need to manage window placement itself, the same way we provide a `terminal` alias that various other scripts use to open a new terminal we can provide a `terminal-vertical` that would do its best effort to create a vertically split terminal, if that is not possible let the user place the window and be done with it.

**Comment by dumblob** [comment]
Speaking of "side by side", please don't forget there is 3-way diff as well as need for arbitrary number of splits (I use it quite often on nowadays wide diplays - I have 4+ splits there in neovim).

**Comment by mawww** [comment]
@dumblob I imagine that encouraging all `terminal-vertical` implementation to split in a consitent way (new terminal on the right of the old one) would make those cases manageable, and when the backend is unable to provide any guarantees, it still becomes the users responsibility to place windows as expected (but the more windows, the more complex it becomes for the user).

**Comment by listentolist** [comment]
For a vimdiff-like diff view it would be nice if it was possible to give the `replace-ranges` highlighter a newline character. That way we can add a newline for every line, that is deleted in the other diff file, so that the files are aligned. So if the first line in the first file is not present in the second file, it would be something like this:
```
set window my_range %val{timestamp} '1.1+0|deleted\n'
```

**Comment by crocket** [comment]
Is there any diff program for kakoune?

**Comment by tototest99** [comment]
https://discuss.kakoune.com/t/compare-selections-with-gnu-diff-utility/1691

**Comment by crocket** [comment]
I noticed that kakoune's `:new` command can be used to implement kakdiff.

**Comment by stacyharper** [comment]
Hey there. I'm sorry to re-open the discussion but I really miss a kakdiff :S

Basically being able to kakdiff two or more files and just seeing line prefixes `+/-/~` would already suits me. I dont ask a scroll synchro, I dont ask smart shortcut. Atm this simplest solution would really help and close this gap in my workflow tools.

Someone got something that do that ?

**Comment by crocket** [comment]
I just use vimdiff for now.

**Comment by Screwtapello** [comment]
When people say "kakdiff", it's not entirely clear what they mean, from just `diff -u foo bar | kak` to a full three- or four-way visual merge tool like [Meld](https://meldmerge.org/) or [kdiff3](http://kdiff3.sourceforge.net/).

One of Kakoune's major design decisions is "Kakoune should not manage the size, position, or layout of Kakoune windows". That means Kakoune works equally well in a terminal-based environment like tmux, a minimalist graphical environment like dwm, or a rich desktop environment like GNOME or KDE, but it also means Kakoune can't reliably do the "two panes side-by-side with a shared scrollbar" view that most visual merge tools require. For anyone who feels that's a minimum required feature, you will probably never have a "kakdiff" tool you like.

At the other extreme, if your use-case is "sometimes `git merge` has a merge-conflict, and I want to be able to resolve it in my editor" then you should be fine: `git merge` writes merge conflict hunks into the file with the conflicts, you can load that up into Kakoune, search for `<<<<` and decide for yourself which parts of the remote branch and the local changes you want to be reflected in the resulting merged file.

It would be reasonable to write a Kakoune command to take a buffer in "unified diff" format, writing the common and `-`-prefixed lines to the `-` file, and the common and `+`-prefixed lines to the `+` file. Then you could diff two files, pipe the result to Kakoune, make certain changes and write them back out, which *might* be useful for certain kinds of merging under certain circumstances, but certainly not all.

**Comment by basbebe** [comment]
For the common git merge, I use kakoune or [kakmerge](https://github.com/lenormf/kakmerge) (although this doesn't seem to work for me sometimes) at the moment.

For situations where I e.g. want to update a config file from a newer default version, I use a combination of [delta](https://github.com/dandavison/delta) and sdiff.

This use case is rather simple and   I would _love_ to use kakoune for it.

It could be covered by what @Screwtapello described.

**Comment by krobelus** [comment]
I have a very rough draft in https://github.com/krobelus/kakoune/commits/kakdiff.
To try it without installing run `make debug=yes` and `PATH=$PWD/src:$PATH contrib/kakdiff old-file new-file`.
(This only works if you don't override the autoload directory; if you do maybe prepend something like `HOME=$PWD`.)

- This shows old/new files in two split windows and adds +/- line flags before added/deleted lines
- `<ret>` scrolls the other window to the matching position (no automatic scrolling)
- `ga` will jump to the `*diff*` buffer, and pressing `<ret>` will make both the split windows jump to the old/new positions.
  -  Whenever you press `<ret>` in either the old/new buffer, the location in the (hidden) `*diff*` buffer will update as well.
- In case you edit the files, the `*diff*` buffer is not updated on the fly (like in vimdiff) but only on a BufWritePost hook.

Let me know if you have suggestions.
(Documentation is missing and there are many unhandled edge cases, but let's figure out the UI first)

**Comment by krobelus** [comment]
BTW a much simpler solution already landed in master recently: simply run `diff -u old-file new-file > my.diff && kak my.diff`
Inside the diff buffer, use `:diff-jump` (bound to `<ret>`) to jump to the corresponding position in the new file and `:diff-jump -` for the old file.

**Comment by stacyharper** [comment]
> When people say "kakdiff", it's not entirely clear what they mean, from just diff -u foo bar | kak to a full three- or four-way visual merge tool like [Meld](https://meldmerge.org/) or [kdiff3](http://kdiff3.sourceforge.net/).

To me kakdiff would be a way to manage file difference in a convenient way with the kakoune paradigms. I dont want a vimdiff iso nor any other solutions.

> At the other extreme, if your use-case is "sometimes git merge …

Ya ideally kakdiff would be my solution when I dont have git :]

The @krobelus solution look great. I think it need a little update cause some methods arent defined ("new-horizontal", "no such client: 'new'") but that shouldnt be hard to do. I may use it, at least temporary.

> BTW a much simpler solution already landed…

I just pulled and build. I dont have those "diff-jump" commands. Maybe I did something wrong o_O But that already would be enough for my cases ! Having a little kakdiff wrapper and a post buffer write hooks would make it perfect

**Comment by krobelus** [comment]
> I think it need a little update cause some methods arent defined ("new-horizontal", "no such client: 'new'")

Whoops sorry, fixed in the latest version.

**Comment by krobelus** [comment]
Added a `kakdiff1` command that does everything in one window (jump back and forth with `<ret>`). It also updates the diff on save (not strictly necessary but it's a simple UI).

**Comment by ismay** [comment]
> BTW a much simpler solution already landed in master recently: simply run `diff -u old-file new-file > my.diff && kak my.diff` Inside the diff buffer, use `:diff-jump` (bound to `<ret>`) to jump to the corresponding position in the new file and `:diff-jump -` for the old file.

Building on this I've created a simple kakdiff function in fish:

```fish
function kakdiff --description "Diff files in kakoune" --argument-names old new
    # Create temporary file for diff
    set -l diff (mktemp --tmpdir="$XDG_RUNTIME_DIR")

    # Set absolute paths so diff-jump works from the temporary file
    set -l old (realpath $old)
    set -l new (realpath $new)

    diff -u $old $new >$diff

    # Exit if diff fails
    if test $status -eq 2
        return 1
    end

    # Update diff when files change
    printf %s\n $argv | entr -nps "diff -u $old $new >$diff" &>/dev/null &
    set -l entr $last_pid

    kak $diff

    # Clean up after editor exits
    kill $entr
    rm $diff
end
```

It uses entr (https://github.com/eradman/entr) to watch the two diffed files for changes, and recreates the diff if either changes. I have this in my `.config/fish/functions` as `kakdiff.fish`. That allows me to run `kakdiff old.file new.file` and see the diff.

As krobelus mentioned, `<ret>` will jump to the new file, and `:diff-jump -" to the old file. There you can make edits, and after saving the diff will be automatically updated. When back at the diff kakoune will prompt you that the diff has changed (if it has).

That way you can easily navigate between diff and compared files without spawning multiple windows. Hope this helps.

**Comment by stacyharper** [comment]
Interesting @ismay +1

For those who want a shell version: https://paste.sr.ht/~stacyharper/1ddbc2c990fd9edfbfd0b340854574560ce3de10

**Comment by ismay** [comment]
Thanks @stacyharper 

One other thing worth mentioning. You can also use the above kakdiff script to diff files owned by root with:

```bash
SUDO_EDITOR=kakdiff sudo -e old.file new.file
```

For this to work, the kakdiff script has to be on your `$PATH`, so it can no longer live in your fish functions. After moving it into my path and making it executable, I've adapted the script slightly as well, to this:

```fish
#!/usr/bin/env fish

if test (count $argv) -lt 2
    echo 'need two files to diff'
    return 1
end

if test (count $argv) -gt 2
    echo 'cannot diff more than two files'
    return 1
end

for file in $argv
    if not test -e $file
        echo "$file does not exist"
        return 1
    end
end

# Create temporary file for diff
set -l diff (mktemp --tmpdir="$XDG_RUNTIME_DIR")

# Set absolute paths so diff-jump works from the temporary file
set -l old (realpath $argv[1])
set -l new (realpath $argv[2])

diff -u $old $new >$diff

# Exit if files are the same
if test $status -eq 0
    echo 'files are identical'
    return 1
end

# Exit if diff fails
if test $status -eq 2
    return 1
end

# Update diff when files change
printf %s\n $argv | entr -nps "diff -u $old $new >$diff" &>/dev/null &
set -l entr $last_pid

kak $diff

# Clean up after editor exits
kill $entr
rm $diff
```

Sorry for the verbose comment. Hope it's useful, feel free to hide if not. See https://man7.org/linux/man-pages/man8/sudoedit.8.html
